### PR TITLE
Reinstate disposition header for email messages

### LINKED ISSUE #25076: Magento 2.3.3 notifications converted to attachments in MS Exchange - Disposition Header
### Preconditions (*)

1. Magento version 2.3.3 (worked fine previously on 2.3.2)
2. Outlook 2010 behind MS Exchange 2010 (don't have later version accessible for immediate testing with this project), or Outlook Web Access, have also confirmed with Mail clients on Mac, iPhone and Android that pull mail from this Exchange server
3. Issue does not appear to affect rendering from Gmail/etc - seems to only be clients that are connecting to MS Exchange 2010 to retrieve the message

### Steps to reproduce (*)

1. Submit an order, send order notification or send shipment notification
2. Receive email using a client/server combination similar to above noted configuration (Outlook/Exchange)

### Expected result (*)

1. Receive an email with the body of the message containing the rendered template for the order notification email

### Actual result (*)

1. Receive an email with a blank body, and the content included as an ATT*-labeled attachment to the email
2. EmailContentAsAttachment
 
### Notes
I understand that this Outlook/Exchange combination is fairly aged, but it's what I'm working with for at least the next couple of months.

I believe this is associated with the changes from #23643, putting the quoted-printable [back] into the emails.

From what I can determine, when a header is rendered, it's processed by a class implementing `Zend\Mail\Header\HeaderInterface`. There is no class registered for the header "Content-Disposition", so it ends up getting rendered through the `Zend\Mail\Header\GenericHeader` class. The `getFieldValue()` method looks like this:

```php
    public function getFieldValue($format = HeaderInterface::FORMAT_RAW)
    {
        if (HeaderInterface::FORMAT_ENCODED === $format) {
            return HeaderWrap::wrap($this->fieldValue, $this);
        }

        return $this->fieldValue;
    }
```

When commenting out the if() block above, the email looks fine when it's received. What's happening is that the Content-Disposition is getting set to "inline", however when rendered it is being translated as quoted-printable and ends up with a value of `=?utf-8?Q?inline?=`, instead of `inline`. The MS Exchange 2010 involved appears to reject or not understand this disposition type and treats it as a generic attachment.

I don't believe that it's necessary/good to have the Content-Disposition header encoded like this. Maybe it is. I'm not completely familiar with the relevant RFCs and it's completely possible that this Exchange server is not handling it correctly. I will be attempting to find a workaround I can implement as a local plugin for now, but wanted to get this out in front of more experienced Magento developers in case the impact is greater than an old and just about out-of-support version of MS Exchange.

### Thank you!
The 2.3.3 rollout has been working pretty well for us so far outside of this issue - all the hard work is much appreciated, you guys rock!

### LINKED ISSUE #29258: Magento does not send emails with Content-Disposition: inline headers anymore
Right from the beginning, Magento has always sent emails with Content-Disposition: inline set as a header. This was the default implementation of MIME encoded emails provided by Zend Framework 1. Magento never needed to do anything as it was handled within ZF1.

When Magento moved its mail handling routines to Zend Framework 2, the content-disposition header, along with other functionality, such as mime encoding text based emails so they can be UTF-8 encoded was lost during the transition (This functionality was previously provided by ZF1 by default, but with ZF2 it was now Magento's responsibility to implement). I patched these missing items back around July 2019 with a series of commits (https://github.com/magento/magento2/commit/fa030f284662ea698ea149b1b5de50cc9ea10e68, https://github.com/magento/magento2/commit/226009e90bba3af2c1d2b0ddb3ad9fc5c2b8f362).

Unbeknown to me at that time, ZF2 contained a bug which caused the Content-Disposition: inline header to be UTF-8 encoded which caused some problems with some mail clients adding the body of the message as an attachment.

Because of this issue, another commit was made (https://github.com/magento/magento2/commit/6976aabdfdab91a9d06e412c2ed619538ed034b6) which removed the Content-Disposition: inline header, as removing the header altogether made the email clients that previously had trouble with the incorrectly UTF-8 encoded header, now work.

Laminas-Mail have now fixed the issue in release v2.11.0 so this can now be looked at again, and the Content-Disposition: inline header reintroduced, as it always has been in previous Magento releases.

See for further discussion

https://github.com/magento/magento2/issues/26913
https://github.com/mageplaza/magento-2-smtp/issues/227
https://github.com/magento/magento2/issues/25076
https://github.com/laminas/laminas-mail/issues/2

### Preconditions (*)
1. 2.4-develop

### Steps to reproduce (*)
1. Configure the store in any way you like and trigger an email to be sent

### Expected result (*)
1. Email should be received and have the "Content-Disposition: inline" header, and it should not be UTF-8 encoded.

### Actual result (*)
1. "Content-Disposition: inline" header is not present at all.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by fredden**
### Description (*)
This pull request reverts the work-around introduced temporarily as part of #25076 and changes the minimum required version of upstream package `laminas/lamnias-mail` to fix the actual problem identified in #25076. External package has already been updated to fix the bug in https://github.com/magento/magento2/commit/5498e7613e51288184f1c1e5a6368544efa4b7bb.

### Fixed Issues (if relevant)
- Fixes https://github.com/magento/magento2/issues/25076
- Fixes https://github.com/magento/magento2/issues/29258
- Not strictly a pull request, but related commit that was added without a pull request: https://github.com/magento/magento2/commit/6976aabdfdab91a9d06e412c2ed619538ed034b6 / see https://github.com/magento/magento2/issues/25076#issuecomment-558339798

### Manual testing scenarios (*)
1. Generate an email from Magento (eg, new user welcome, sale order success)
2. Observe headers in email

### Questions or comments
*None*

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by fredden** [comment]
@magento run all tests

File: composer.json [hunk 1]
```diff
@@ -47,7 +47,7 @@
         "laminas/laminas-http": "^2.6.0",
         "laminas/laminas-i18n": "^2.7.3",
         "laminas/laminas-json": "^3.2.0",
-        "laminas/laminas-mail": "^2.9.0",
+        "laminas/laminas-mail": "^2.11.0",
```
Comment (ihor-sviziev) [inline composer.json]: Hi @fredden,
Could you update laminas/laminas-mail in the Magento framework to fix static test failure:
https://github.com/magento/magento2/blob/2cbb4e6557139340cbeb5c0773fb078f41cfcc6e/lib/internal/Magento/Framework/composer.json#L31
Thank you!

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by fredden** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
In general, looks good, but now the conflict appeared, also there still static test failures caused by different dependencies in the b2b and/or ee versions. Could you check that?

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run static tests

**Comment by fredden** [comment]
@magento run Static Tests

**Comment by fredden** [comment]
@magento run Static Tests

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests, WebApi Tests

**Comment by ihor-sviziev** [comment]
@magento run WebAPI Tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9023](https://jira.corp.magento.com/browse/ENGCOM-9023) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by ihor-sviziev** [comment]
@magento run Integration Tests

**Comment by engcom-Oscar** [comment]
:heavy_check_mark: QA passed

Was checked case from Manual testing scenarios 

:x: **Before:**
"Content-Disposition: inline" header is not present at all.

:heavy_check_mark: **After:**
Screenshot from 2021-04-19 13-38-54

**Comment by ihor-sviziev** [comment]
Moved back to the correct status.

**Comment by ihor-sviziev** [comment]
Moving back to "accept" status. It looks like it was moved by some mistake.

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
Moving back to "accept" status. It looks like it was moved by some mistake.

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE

**Comment by ihor-sviziev** [comment]
@sidolov @sivaschenko @gabrieldagama,
The bot is constantly moving this PR from "accept" to "ready for testing". Could you take a look into it?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@fredden, could you resolve the merge conflict? 
That's the reason this PR isn't going forward.

Thank you!

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by ihor-sviziev** [comment]
The changes were only in the hash parameter of composer.lock file, so it won’t affect the functionality. Moving back to accept status

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9023](https://jira.corp.magento.com/browse/ENGCOM-9023) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Hi @ihor-sviziev , @fredden 
Thanks for the contribution!
I tried to reproduce utilizing outlook and could not successfully reproduce the case.
Trying to work on MS Exchange server, and will post here for any further updates!

**Comment by engcom-Alfa** [comment]
Hi @fredden, @ihor-sviziev 
Thanks for the collaboration & contribution!

:x: QA not Passed

Disposition header is missing in the welcome/sales notification mails

**Manual Scenario Steps**

1. Set the mail magepal gmail smtp in Magento installed instance.
2. Do create a new customer/ place an order.
3. 3. Check the mail body in the web-outlook.


**:heavy_check_mark: Expected result After Fix** 
- Content-disposition should be included


**:x: Actual result After Fix** 
- Content-disposition still not included
- image

Additionally I am unable to test with MS Exchange Server, but anyhow, I am unable to see the Content-Disposition in the mail DOM.

**Comment by ihor-sviziev** [comment]
@engcom-Alfa, are you sure that this too shows the original email headers?

**Comment by engcom-Alfa** [comment]
Hi @ihor-sviziev Thanks for discussing with me more on this issue. That helped me in taking decision!

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Should have a Magento instance installed with mailing configured

**Manual testing scenario:**

1. Do create a new customer/ place an order.

2. Check the mail body in the web-outlook (show original).


**Before:** :heavy_multiplication_x: Content-disposition was missing

**After:** :heavy_check_mark:  Content-disposition is visible as shown in below screenshot.
- image


There is no other additional testing is required and there is no impact on any other feature as well.

**Comment by sdzhepa** [comment]
@magnto run all tests

**Comment by sidolov** [comment]
@magnto run all tests

**Comment by sidolov** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@sidolov, could you move this PR to "accept" again?

**Review by sdzhepa — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B

**Comment by sidolov** [comment]
@fredden could you please resolve merge conflict? Thanks!

**Comment by fredden** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests CE, Functional Tests EE, Static Tests, Unit Tests

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
@fredden thanks for your contribution and collaboration!
The testing has be done as per [above-comment](https://github.com/magento/magento2/pull/32720#issuecomment-915123737).
There are some merge conflicts needs to be corrected. Hence moving it to _Changes Requested_
Kind request to resolve conflicts; Thanks in advance!

**Comment by ihor-sviziev** [comment]
@andrewbess any thoughts on why static tests are failing?

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@fredden
FYI the changes from this PR were already merged to 2.4-develop branch in https://github.com/magento/magento2/commit/4a125b0b887e7f2cf380dbc3eae122eb736447eb except the latest merge commit. As a result, the pull request wasn't marked as merged.
@sidolov will work on it to mark it as fully merged

**Pull request merged**